staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go

```diff
@@ -463,7 +471,10 @@ func ExpandPathsToFileVisitors(mapper *mapper, paths string, recursive bool, ext
 		if path != paths && ignoreFile(path, extensions) {
 			return nil
 		}
-
+		if strings.HasSuffix(path, "kustomization.yaml") {
```

Comment:
why not `filepath.Base(path) == "kustomization.yaml"`

Reply:
I thought the intent of cli-runtime was to be generic function for things consuming resource builder functionality. Hard-coding this seems extremely specific and not very scalable.

Can we pull this out to a package var or registration method that kubectl could inject special handling for `kustomization.yaml` into, rather than making all cli-runtime consumers pick up `kustomize`?

Reply:
It sounds to me like there are two possible approaches to the kustomize implementation:

1. kustomize remains unique to kubectl, it lives in kubectl, and any extra behavior it needs to introduce to the resource-builder (or any other generic utilities used to consume resources) is done from a "initialization" function within kubectl.

2. kustomize becomes a part of the overall cli ecosystem, we keep it hardcoded within genericclioptions, and `kubectl` along with third-party clients and plugins all inherit this new capability of handling "kustomize" directories

I would be fine with either of these two approaches, however I do see value in propagating this behavior to consumers of genericclioptions, as it would maintain this similar behavior with kubectl.

cc @soltysh @smarterclayton @deads2k for additional thoughts on this

Reply:
@pwittrock  I'll update the statement. @smarterclayton  We can change the hard coded `kustomization.yaml` by a var or registration method. That will require the change from the kustomize side. Can we do it in a following PR?

@juanvallejo I prefer it becomes a part of the overall cli ecosystem so that other clients can also use kustomize by default.

Reply:
So Juan and I discussed this in depth and here's our proposed solution. We picked a middle-ground solution from what he proposed in the previous comment. This means that cliruntime will have extension point (like Jordan proposes) where an author would be able to include additional functionality (eg. `cliruntime.Addons["kustomization.yaml"] = kustomizationHandler()`). Kustomize would be such an extensions, our first one. It would entirely live inside its current repo (https://github.com/kubernetes-sigs/kustomize). 
This way kubectl (as its first consumer) would vendor both cliruntime and kustomize and initilize kustomize as an addon for cliruntime. This would allow few things:

1. keep cliruntime generic and very minimal
2. provide extensions points for cliruntime (builder to start with)
3. give cliruntime consumers bigger flexibility over what they're getting.

@pwittrock @liggitt @juanvallejo @deads2k @smarterclayton what do you think
@Liujingfang1 I'm sorry we're changing the design one more time :relaxed:

Reply:
@soltysh Didn't see your comment, responses inlined.

> 1. keep cliruntime generic and very minimal

This also makes the default behavior for clients inconsistent with the default behavior for kubectl commands such as apply, get, describe.  May be it would be best to have a simple and modular architecture by making all of the Resource Builder targets be plugins.

> 2. provide extensions points for cliruntime (builder to start with)

I'd like to see more discussion about this and what the interfaces might look like before committing to it.

> give cliruntime consumers bigger flexibility over what they're getting.

This can be done more simply by making it opt-out

Reply:
For context, from the cli-runtime readme:

> [Purpose](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cli-runtime): 
> This library is a shared dependency for clients to work with Kubernetes API
> infrastructure which allows to maintain kubectl compatible behavior.
> Its first consumer is k8s.io/kubectl.

To be useful in a k8s context, cli-runtime must be cognizant of API [guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md) and recognize (depend on) things like `TypeMeta` and `ObjectMeta`, and be able to wrangle YAML/JSON, rest, patching, etc, even as it avoids direct dependence on the API types.

kustomize was built with the same goals and restrictions.  It was developed as a standalone project (outside k/k) as an expediency to release faster, get feedback faster, prove utility and to avoid making the kubectl extraction project more difficult.  It was always intended to become part of the builder, to enhance the CLI’s ability to manipulate k8s-style objects in multiple commands and contexts. 

It’s a cross-cutting enhancement to manipulate k8s YAML meant to collectively address [old kubectl bugs mentioned in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/0008-kustomize.md#long-standing-issues).  We should fix these issues for all runtime consumers, not just kubectl users.

So please accept this PR (more or less as is, modulo how the file name is handled), and if there's some outcry from CLI users who don't want these fixes/features, we can factor it out of the runtime then.